### ISSUE TITLE
REST API For OSPOS 

### ISSUE THREAD

**Issue by jefenga** [opensourcepos/opensourcepos#2463]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.2.3
- OSPOS git commit hash is: 
- PHP version is: 7.2
- MySQL or MariaDB version is: 5.7
- OS and version is: ubuntu 19.04
- WebServer is: apache
- Selected language is: 
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: lamp

### Issue / Bug  / Question / New Feature

Please write your issue here. If a bug, please make sure to provide as much information as possible including configuration settings (e.g. Decimals set, Tax mode), language and steps to reproduce the bug.

Hello,
Am using OSPOS for my sales field force team, so we are develop an android app that would be sending and receive data from ospos db, we are using android app because it allow us to use even if we are offline. my question is the a way or may be in a future  ospos would do enhancement for rest api intergration

Thank you 
Apologies for my English

**Comment by daN4cat** [comment]
Yes, REST API can be added but it depends on which functionality you are after: sales register, inventory, customer list, reporting, etc..

**Comment by jefenga** [comment]
@daN4cat Thanks for quick reply. what framework do you suggest we could go with? am after sales register, inventory customers and suppliers

**Comment by objecttothis** [comment]
I am currently working on the framework for 3rd party integrations which will provide code-space within the OSPOS project to write 3rd party integrations which can include REST API integrations.  Currently the space exists for settings of each integration (API URL, API key, integration-specific settings, etc) and we are working out the best way to implement the "event dispatcher" portion of the 3rd party integrations.  Hooks don't seem to be a good fit for event dispatching because of the inability to pass appropriate data to the hook, so we may end up writing a library to do this functionality.

**Comment by waploaj** [comment]
@objecttothis That is great, i would love to contribute on that

**Comment by objecttothis** [comment]
> @objecttothis That is great, i would love to contribute on that

@waploaj This is good.  To be clear, API related integrations have two parts of them.  Client and Server.  The client-side 3rd party integrations added into OSPOS would be specific to the integration in question.  For example, the CLCdesq integration I am writing code for makes API calls to an API server offsite.

The server-side API aspect of OSPOS would need to be written so that all integrations could use it rather than a separate REST API server for each.

The first step with this is to come up with the spec for the REST API 'listener'.  We need to document well each element well.  I'm assuming that aside from JSON encoded, that an API key would be issued by OSPOS and required to be sent along with the message to the API endpoint.

Whether we implement a single API endpoint for all functions or multiple endpoints for different functionality as @daN4cat pointed out, will be something for discussion as well.

**Comment by discapacidad5** [comment]
Hi I do not speak English, I am using the Google translator but I am also interested in this AAPI
I would like that when you sell an item the system sends an Api Rest to an external system
the idea would be that the program has a space in which you place your reset code of your external system and in each article you place the variable that is filled in the code in this way it would be an app and open that could be configured for any program external
  for example in my case I want to use it with a program called radiusdesk that is for internet sale if someone buys a ticket from the internet the system automatically sends the recipe to radiudesk to recharge the amount

**Comment by daN4cat** [comment]
@objecttothis is working on hooks to call external APIs.
It's part of a development branch.

**Comment by objecttothis** [comment]
@discapacidad5 once the 3rd party integration feature is released, you will still need to develop the integration that talks to the API that you are wanting to communicate with.  I am developing one for a CMS that we use, but it's specific to that API.

**Comment by discapacidad5** [comment]
> @discapacidad5 once the 3rd party integration feature is released, you will still need to develop the integration that talks to the API that you are wanting to communicate with. I am developing one for a CMS that we use, but it's specific to that API.

Not necessarily
Because if for example in the OSPOS in each article I put an additional field that each person can name and that field becomes a variable
 then in settings created an empty field to paste an Api Rest & with the parameters as I want the RESET to be sent
 those parameters will collect the variables that were saved in each product then in the pay button placed the shipment of that API
I am making it an open program and can be used with any API of any existing program that supports API since each person would configure the API according to their need

the only limitation that I see this solution is that you can only buy one product at a time through the API
if you want to buy two different products you would have to make two separate payments

At least I could configure you to make a shipment for each product

I do not speak English I am using the Google translator I do not know if my comment was understood If it is understandable Sorry for the bad English of the translator's product

Here is an example as an idea.


```
What would my custom file be like I want with the parameters that I want to be pasted in a field in settings.
/////////////////////////////////////////////////////////////////
// This would be the modifiable code field to suit your needs.

$active         = 'active';
$cap_data       = 'soft';
$language       = '4_4';
$parent_id      = 0;
$profile_id     = 9;
$realm_id       = 34;
$token          = 'xxxxxxx-xxxx-xxxx-xxx-xxxxx';

//////////////////////////////////////////////////////////////
// In this fault would go the fields that we want to capture from the system to send it by means of variable.

// these are the data that would be taken from the fields of
$username       =  me.Username; // Here the name of the custom field we want to capture.
$quantity       =  me.quantity; // here I wanted the quantity field or the amount field depending on the needs we want to send to the REST
 
$url            = 'http://my url/add.json'; // Acquire the address where respect would be sent when the payment is made.

////////////////////////////////////////////////////////////
//  This field would also be the user modifiable tablet.


// The data to send to the API
$postData = array(
    'active'        => $active,
    'cap_data'      => $cap_data,
    'language'      => $language,
    'parent_id'     => $parent_id,
    'profile_id'    => $profile_id,
    'realm_id'      => $realm_id,
    'token'         => $token,
    'username'      => $username,
    'quantity'      => $quantity
);
 
////////////////////////////////////////////////////////////
// this part of the code would be internal and not modifiable by the user is the one that defines the sending of the array as Json

// Setup cURL
$ch = curl_init($url);
curl_setopt_array($ch, array(
 
    CURLOPT_POST            => TRUE,
    CURLOPT_RETURNTRANSFER  => TRUE,
    CURLOPT_HTTPHEADER => array(
        'Content-Type: application/json'
    ),
    CURLOPT_POSTFIELDS => json_encode($postData)
));
 
// This code should be called when processing payment or registration as appropriate.

// Send the request
$response = curl_exec($ch);
 
// Check for errors
if($response === FALSE){
    die(curl_error($ch));
}
 
// Decode the response
$responseData = json_decode($response, TRUE);
print_r($responseData);
 

```

**Comment by waploaj** [comment]
> Hi I do not speak English, I am using the Google translator but I am also interested in this AAPI
> I would like that when you sell an item the system sends an Api Rest to an external system
> the idea would be that the program has a space in which you place your reset code of your external system and in each article you place the variable that is filled in the code in this way it would be an app and open that could be configured for any program external
> for example in my case I want to use it with a program called radiusdesk that is for internet sale if someone buys a ticket from the internet the system automatically sends the recipe to radiudesk to recharge the amount

I have create a custom apis for receiving,sales,customer,items and supplier module , that communicate with our android app. But we have customise it to fit our business needs.i could share the project so you can tweak it to meet you requirements.

**Comment by jekkos** [comment]
@waploaj sure it would be nice to see what you made there and whether it is usable as a generic version. If you did a generic API for all these modules then I hope you could take the time to submit a PR so we can have a look there?

**Comment by jekkos** [comment]
if you think about it , one can probably just use the standard controllers as most of the responses are JSON anway (due to bootstrap-tables).

The only annoyance is that you need a csrf token and session id to be able to call these methods (search, get_row, update).. but beside that it might only be a matter of documenting those methods

**Comment by dme1sc** [comment]
Where can I find any established API documentation for ospos?  I'm not finding much or I'm not looking in the right places.  I'm looking to build two custom payment tender buttons and was hoping to follow in the footsteps of giants instead of inventing the wheel for this stuff on my own.

**Comment by daN4cat** [comment]
Because there is none. We haven't implemented any REST API yet.

**Comment by objecttothis** [comment]
@dme1sc It's on the roadmap for changes.  Right now I am working on the CI4 upgrade which will pave the way for me to do events which in turn paves the way for 3rd party integrations.  After that I will be working on two third party integrations that we need for our operation (API calls to interact with the government cash registers, and with our in-house CMS for our website).  Once those are all done then I will be free to work on REST API endpoints for OSPOS.  We will need to do some planning together to see the most useful API endpoints.  If anyone wants to speed the process along, they can help me with the CI4 upgrade.

**Comment by jekkos** [comment]
@dme1sc I think most controllers actually do have GET/UPDATE/DELETE/INSERT methods
All of them return JSON output currently. Maybe the only annoyance is that you need a cookie and CSRF token for authentication and authorization.

The cookie needs to be obtained once after login. Then after the CSRF token needs to be parsed in some way and sent back with any POST request.

**Comment by jekkos** [comment]
There is however no documentation on this ... so basically it will need to be reverse engineered by reading the code, which I suppose is not really ideal.

**Comment by objecttothis** [comment]
I think we need to introduce an API key system... Not sure yet how that would interact with the current csrf token, but it would eliminate the need to login per se.

**Comment by dme1sc** [comment]
You mention the need for CSRF token - is there documentation somewhere to get this generated?

**Comment by dme1sc** [comment]
Another thought here is that I am running this on a raspi on tor and ospos is never meant to be accessed on the clearnet.  The app I am integrating will be localhost - understanding there is an inherent risk if I had this installed on an openly accessible domain - can't I just turn CSRF off for this install?

**Comment by dme1sc** [comment]
The ignore URI ended up working great for our needs.

**Comment by jekkos** [comment]
CSRF should only be applied to endpoints that mutate application state. In this case it is probably fine to disable it on the GET read endpoint using the ignore list indeed.

**Comment by objecttothis** [comment]
> CSRF should only be applied to endpoints that mutate application state. In this case it is probably fine to disable it on the GET read endpoint using the ignore list indeed.

Agreed but we still need to add functionality to issue API keys and add grants to those keys so that not everyone who sends a GET request gets a reply.

**Comment by DamianoSilverhand** [comment]
@waploaj I am trying to implement a custom REST Api for items, would really love to see how you achieved it , that is if you don't mind sharing the api part of your project.

**Comment by waploaj** [comment]
@DamianoSilverhand you can find it here [Link](https://github.com/waploaj/ospos-api.git) on main branch

**Comment by emushene** [comment]
> @DamianoSilverhand you can find it here [Link](https://github.com/waploaj/ospos-api.git) on master branch. It's a custom API, so you'll find other modules on top of OSPOS that are unrelated

@waploaj 

are you running these in your kiboApi/kibo?

**Comment by geeknation** [comment]
@daN4cat are there plans to build the api any time soon?

**Comment by objecttothis** [comment]
> @daN4cat are there plans to build the api any time soon?

There are two separate things meant when referring to APIs.
- API Endpoints in OSPOS to receive requests then do work/respond accordingly.
- Third-party interfaces to make API calls to a external API endpoints.

The latter will need to wait on #3271 since CI4 has the events capabilities.  @jefenga please clarify which one you are requesting?

The former requires endpoints inside OSPOS and will probably be as follows (please respond with others that are needed, but I missed)
- Item Push (Create, Read, Update, Delete)
- Sale Push (Create, Read)... is Update/Delete needed here?
- Customer Push (Create, Read, Update, Delete)

Then later other endpoints can be added as needed.

**Comment by waploaj** [comment]
> > @DamianoSilverhand you can find it here [Link](https://github.com/waploaj/ospos-api.git) on master branch. It's a custom API, so you'll find other modules on top of OSPOS that are unrelated
> 
> @waploaj
> 
> are you running these in your kiboApi/kibo?

no, I am using it in field force Android application or I understand you incorrectly

**Comment by waploaj** [comment]
> > @DamianoSilverhand you can find it here [Link](https://github.com/waploaj/ospos-api.git) on master branch. It's a custom API, so you'll find other modules on top of OSPOS that are unrelated
> 
> @waploaj
> 
> are you running these in your kiboApi/kibo?

No, it is located in ospos_api in master branch, I have provided a link. did you get it?.
